File: Python/TicTacToe/TicTacToe.py
import random


# define global variables
board = [[" ", " ", " "], [" ", " ", " "], [" ", " ", " "]]
userTeam = "-"
cpuTeam = "-"
winner = "-"

def main():
    # Initialise Game 
    turn = chooseTeam()

    print("===============")
    while not isWinner():
        turn = turn + 1
        nextTurn(turn)
        printBoard()
        print("===============")
    
    print(getWinner())

def chooseTeam():
    while True:
        team = input("Enter Team (X or O): ")
    
        global userTeam
        global cpuTeam

        if team == "X":
            userTeam = "X"
            cpuTeam = "O"
            return -1
        elif team == "O":
            userTeam = "O"
            cpuTeam = "X"
            return 0
        else:
            print("Wrong Input! Try Again")
    
def nextTurn(turn):

    if turn % 2 == 0:
        userTurn()

    else:
        cpuTurn()


def userTurn():

    while True:
        try:
            result = input("Enter Position: ")

            if applyIfValid(result, userTeam):
                return
        except:
            print("Wrong Input! Try Again")

def cpuTurn():

    print("CPU's Response:")

    while True:
        result = getRandomNumberInRange(1, 9)

        if applyIfValid(result, cpuTeam):
            return

def applyIfValid(pos, team):

    pos = int(pos)
    count = 0
    for row in range(0, 3, 1):

        for column in range(0, 3, 1):

            count = count + 1

            if count == pos and board[row][column] == " ":
                board[row][column] = team
                return True
            
    return False

def getWinner():

    if winner == userTeam:
        return "The User Wins!!!"
    
    elif winner == cpuTeam:
        return "The CPU Wins!!!"
    
    else:
        return "There is no Winner!"

def isWinner():

    if checkRows() or checkColumns() or checkDiagonals():
        return True

    if isBoardFull():
        winner = "-"
        return True
    
    return False

def checkRows():

    for i in range (0, 3, 1):

        if board[i][0] == board[i][1] and board[i][1] == board[i][2] and not board[i][0] == " ":
            winner = board[i][0]
            return True
        
    return False

def checkColumns():

    for i in range (0, 3, 1):

        if board[0][i] == board[1][i] and board[1][i] == board[2][i] and not board[0][i] == " ":
            winner = board[0][i]
            return True
        
    return False

def checkDiagonals():

    if board[0][0] == board[1][1] and board[1][1] == board[2][2] and not board[0][0] == " ":
        winner = board[0][0]
        return True
    
    elif board[0][2] == board[1][1] and board[1][1] == board[2][0] and not board[0][2] == " ":
        winner = board[0][2]
        return True
    
    return False

def isBoardFull():

    for i in range (0, 3, 1):
        for j in range (0, 3, 1):

            if board[i][j] == " ":
                return False
            
    return True

def printBoard():

    print(board[0][0]+" | "+board[0][1]+" | "+board[0][2])
    print("--+---+---")
    print(board[1][0]+" | "+board[1][1]+" | "+board[1][2])
    print("--+---+---")
    print(board[2][0]+" | "+board[2][1]+" | "+board[2][2])

def getRandomNumberInRange(min, max):

    if min >= max:
        raise ValueError("Minimum value should be less than or equal to the maximum value.")

    return random.randint(min, max)
    


if __name__ == "__main__":
    main()

File: Ruby/TicTacToe/TicTacToe.rb

# define global variables
$board = [[" ", " ", " "], [" ", " ", " "], [" ", " ", " "]]
$user_team = "-"
$cpu_eam = "-"
$winner = "-"

def main
	# Initialise Game
	turn = choose_team

	puts "==============="
	while not is_winner
		turn = turn + 1
		next_turn(turn)
		print_board
		puts "==============="
	end

	puts get_winner
end

def choose_team

	loop do
		begin
			puts "Enter Team (X or O): "
			team = gets.chomp

			if team == "X"
				user_team = "X"
				cpu_team = "O"
				return -1
			elsif team == "O"
				user_team = "O"
				cpu_team = "X"
				return 0
			else
				print("Wrong Input! Try Again")
			end
		end
	end
end

def next_turn(turn)
	if turn % 2 == 0
		user_turn
  	else
		cpu_turn
	end
end

def user_turn
  	loop do
		begin
			print 'Enter Position: '
			string_pos = gets.chomp
			result = Integer(string_pos)
			return if apply_if_valid(result, user_team)

			rescue ArgumentError
			puts 'Wrong Input! Try Again'
		end
	end
end

def cpu_turn
	puts "CPU's Response:"
	loop do
		result = rand(1..9)
		return if apply_if_valid(result, cpu_team)
	end
end

def apply_if_valid(pos, team)
	# Map number pad positions to Tic Tac Toe board positions
	position_mapping = [
		[7, 8, 9],
		[4, 5, 6],
		[1, 2, 3]
	]

	(0..2).each do |row|
		(0..2).each do |column|
			if position_mapping[row][column] == pos && board[row][column] == ' '
				board[row][column] = team
				return true
			end
		end
	end
	false
end

def get_winner
	if winner == user_team
		'The User Wins!!!'
	elsif winner == cpu_team
		'The CPU Wins!!!'
	else
		'There is no Winner!'
	end
end


def is_winner
	if winner == user_team
		return "The User Wins!!!"

	elsif winner == cpu_team
		return "The CPU Wins!!!"

	else
		return "There is no Winner!"
	end
end

def check_rows
	# TODO
end

def check_columns
	# TODO
end

def check_diagonals
	# TODO
end

def print_board

	puts board[0][0]+" | "+board[0][1]+" | "+board[0][2]
	puts "--+---+---"
	puts board[1][0]+" | "+board[1][1]+" | "+board[1][2]
	puts "--+---+---"
	puts board[2][0]+" | "+board[2][1]+" | "+board[2][2]

end

# Execute main function
main
